[BUGFIX release] Default {{each}} key to @guid or @item based on type.

rwjblue opened the pull request:
This PR uses `@item` as the key when the item being iterated is a string or number, and uses `@guid` when it is anything else.

This is an attempt at making a better default, to prevent users from having to specify a `key=` to each `{{each}}` invocation.

It is absolutely possible, that we will want to revisit this and make specifying `key` required in the future, but it seems that we should attempt at good defaults before making that decision.

rwjblue committed:
[BUGFIX release] Default {{each}} key to @guid or @item based on type.

mmun commented on packages/ember-htmlbars/lib/utils/decode-each-key.js:
```diff
@@ -19,8 +19,13 @@ export default function decodeEachKey(item, keyPath, index) {
     } else {
-      Ember.warn('Using `{{each}}` without specifying a key can lead to unusual behavior.  Please specify a `key` that identifies a unique value on each item being iterated. E.g. `{{each model key="@guid" as |item|}}`.');
-      key = index;
+      let type = typeOf(item);
```
Why not just `typeof`?

stefanpenner replied:
yes, lets just use JS typeof here

mmun commented:
Looks good :+1:

JKGisMe commented:
yay :+1:

rwjblue commented:
Update:

* Use normal `typeof` (instead of `Ember.typeOf`).
* Added `@detect` as a new type (makes it much easier to document the default)
* Added API docs saying what the default is.

mixonic commented:
I am +1. I've also heard many WAT about the current behavior

rwjblue commented:
Updated (hopefully for the last time) to deprecate `@guid` and `@item` in favor of adding the default of `@identity`.

funtusov commented:
It seems that this commit didn't get into the 1.13.1 tag

sandstrom commented:
Thanks!! :boat:
